### ISSUE TITLE
xcape features (double bind modifiers)

### ISSUE THREAD

**Issue by Eisfreak7** [SirCmpwn/sway#719]
Hi,
For me currently the main thing that stops me from giving wayland a try, is missing xcape support.
Now https://github.com/alols/xcape/issues/67 is open for a while, and it doesn't seem like its going to come anytime soon or even is possible.
Is it correct, that the windowmanager would have to implement these features?
If so, have you considered implementing them?

My specific use-cas is:
- Bind Caps to ctrl when I hold it, escape else
- Bind Tab to alt-gr when I hold it, Tab else

**Comment by SirCmpwn** [comment]
I think you should pitch this to libinput.

**Comment by Eisfreak7** [comment]
Thank you, [I did](https://bugs.freedesktop.org/show_bug.cgi?id=96621).

**Comment by Eisfreak7** [comment]
They [won't fix](https://bugs.freedesktop.org/show_bug.cgi?id=96621), apparently its too high-level for libinput.

**Comment by SirCmpwn** [comment]
Ugh. I'll reopen this but I'm not sure how I want it to take shape in sway.

**Comment by Eisfreak7** [comment]
Thanks
It wouldn't really be a good solution if every compositor had to implement this (although I think that especially people using tiling window managers may profit from this).
But if its the only one, I'm glad that you at least consider it.

**Comment by progandy** [comment]
Is this even possible in sway without changing wlc? I think the whole keymapping is hidden and not accesible.

**Comment by myfreeweb** [comment]
Looks like this might be easier with wlroots.

Plus, here's an idea! Since every problem in Wayland is solved by inventing a protocol…

We need one that's kinda similar to [xwayland-keyboard-grab](https://github.com/wayland-project/wayland-protocols/blob/master/unstable/xwayland-keyboard-grab/xwayland-keyboard-grab-unstable-v1.xml), but:

- obviously, not an exclusive grab — keypresses must be copied to the client;
- the client should provide the set of keys it's interested in — the compositor can implement security policies based on that. e.g. sway can have two different `permit` levels, `watch-modifier-keys` and `watch-all-keys`, or just allow everyone to grab modifier keys (they're not *that* interesting after all) and require `permit watch-keys` for a keylogger;
- and of course, the interface should also include a request that injects keypresses back.

Example flow:

- the hypothetical `wcape` or `waycape` or whatever requests `watch_keypresses(wl_seat, ['Shift_L', 'Shift_R', 'Control'])`
- the user presses and releases Control in a terminal window
- the compositor sends an event for `Control` to the terminal (which ignores it), but also sends that event to `wcape` because it's registered for that key
- `wcape` reads the event and requests `inject_keypress('Escape')`
- the compositor sends an event for `Escape` to the terminal

Bonus idea: clients can request a "secure mode" like in macOS, so when you're entering a password, events would not be copied to `wcape`. (Not super necessary when only giving access to modifier keys, but would be great for, like, running a keylogger for typing statistics I guess :D)

**Comment by Timidger** [comment]
Seems annoying to have to define a whole protocol for this...but it _is_ the most compositor agnostic option :\. 

Unless there was some other layer above libinput that's implemented that can implement this and all the compositors start using that. Though that seems mostly unnecessary and even less likely than all the compositors using the same protocol.

**Comment by myfreeweb** [comment]
Yeah, just setting xcape style key presses is simple enough to do directly in compositors, but this protocol allows for more stuff, such as [TextExpander](https://smilesoftware.com/TextExpander)-like apps, typing statistics, etc.

**Comment by myfreeweb** [comment]
What do y'all think about [this protocol](https://gist.github.com/myfreeweb/7c656d535ae1c5a1336f29d2c1473726)?

**Comment by SirCmpwn** [comment]
-1. This is not something I'm interested in doing with a protocol.

**Comment by myfreeweb** [comment]
So it won't be possible to do UI automation/testing/macros (java.awt.Robot, Sikuli, `xdotool` etc.), text snippet expansion, keyboard typing statistics, etc. under Sway?

I guess I'll have to do this on evdev level then. It's possible to listen directly on evdevs and create a fake keyboard (`uinput` on Linux, `cuse` on FreeBSD) to inject events. I believe that a Wayland protocol would've been a much cleaner and better solution >_<

**Comment by SirCmpwn** [comment]
I don't think that any of those use-cases justify putting in what is *literally* a keylogger protocol extension.

**Comment by myfreeweb** [comment]
What if I want a literal keylogger to analyze my typing speed?

These features are useful. Global input access shouldn't be completely killed, it should be available to privileged applications. macOS uses accessibility APIs for this, you have to tick a checkbox in Preferences to allow an app to do this. Sway has IPC security already, which is pretty similar! `permit /usr/local/bin/xcape keylogging`

**Comment by SirCmpwn** [comment]
Then you should patch sway.

**Comment by myfreeweb** [comment]
What do you mean by "patch"? "maintain your own fork"? :)

Actually, I'll probably do the evdev-level thing because it will work with GNOME, KDE, etc. out of the box. Would help with https://github.com/bulletmark/libinput-gestures/issues/27 as well.

**Comment by SirCmpwn** [comment]
No, I mean patch. I run sway with some private patches myself. Patching software is not a big deal. Just do it until you get bored of gathering stats on your typing habits.

**Comment by oblitum** [comment]
I'm author of this, may be interesting on this topic:

- https://aur.archlinux.org/packages/interception-tools

**Comment by SirCmpwn** [comment]
That's pretty awesome!

**Comment by myfreeweb** [comment]
So, seems like the list of solutions is here: https://github.com/alols/xcape/issues/67

My solution is [evscript](https://github.com/myfreeweb/evscript), it runs [Dyon](https://github.com/PistonDevelopers/dyon) scripts with access to evdev/uinput.

**Comment by emersion** [comment]
Related: https://github.com/swaywm/sway/issues/1779

Note that you can get keyboard identifiers with `swaymsg -t get_inputs`.

**Comment by avrelaun** [comment]
I'm just going to drop my 2 cents here:

**Comment by edrex** [comment]
@avrelaun if you just want to map caps to escape (rather than conditionally to ctrl or escape depending on whether chorded as caps2esc does) you can do that with a hwdb rule as in https://github.com/edrex/etc/blob/master/udev/hwdb.d/90-custom-keyboards.hwdb.

**Comment by emersion** [comment]
> if you just want to map caps to escape (rather than conditionally to ctrl or escape depending on whether chorded as caps2esc does)

The preferred solution is to use the xkeyboard-config option.

**Comment by oblitum** [comment]
For chording keys I think an external tool is preferred over something builtin. It doesn't look like a task for sway. For example, caps2esc has to manipulate key-down and key-up in a specific way to make chording natural and not cause problems like missing keystrokes.

**Comment by avrelaun** [comment]
@edrex I need the exact behaviour with both ctrl and escape depending on the case.

**Comment by jcalonso** [comment]
I came here because I am looking for a way to replace this xcape config:
```
xcape -e 'Shift_L=Shift_L|parenleft;Shift_R=Shift_R|parenright'
```
Any ideas if this is possible to accomplish with Sway? or other Wayland compatible tools?

**Comment by zsugabubus** [comment]
###### Sorry for being off-topic, but if somebody has the same issue and it's the only thing that holds one back to switch, it can be super useful.

@jcalonso, not with (stable) Sway but at a lower level you can achieve it with the above mentioned Interception Tools (by @oblitum). If you could make that work and you don't want to reinvent the wheel you can use [my general purpose plugin](https://github.com/zsugabubus/interception-k2k) for a quick solution.
```c
/* tap-rules.h.in */
{ .base_key   = KEY_LEFTSHIFT, /* change this key */
  .tap_key    = KEY_KPLEFTPAREN, /* what to do when pressed alone/pressed once */
  .repeat_key = KEY_KPLEFTPAREN, /* what to do when pressed for longer time (just remove it if you find it useless) */
  .hold_key   = KEY_LEFTSHIFT /* behave as this key immediately as other keys are pressed */  },
/* similar for right shift... */
``` 

@timokau, your config: 
```c
/* tap-rules.h.in */
{ .base_key = KEY_CAPSLOCK, .tap_key = KEY_ESC, .repeat_key = KEY_LEFTCTRL, .hold_key = KEY_LEFTCTRL },
{ .base_key = KEY_TAB, .tap_key = KEY_TAB, .hold_key = KEY_RIGHTALT/*altgr*/ },
```

**Comment by spectras** [comment]
So, what is the current state of key event copying?

I am the maintainer of https://github.com/keyleds/keyleds. It implements visual feedback on gaming keyboards that have per-key lights, triggering animations to respond to key events.

How would I go about supporting sway/wlroot/wayland, whichever is the right level?

**Comment by emersion** [comment]
>How would I go about supporting sway/wlroot/wayland, whichever is the right level?

We've already said we're not interested in a keylogging protocol. If you want to do this, read events with libinput, like e.g. [wshowkeys](https://git.sr.ht/~sircmpwn/wshowkeys). This requires a privileged process, because regular processes shouldn't have a way to read all key events.

**Comment by spectras** [comment]
Let me get this straight:
- The recommended way to have a secure environment is to force complex applications that interact with the desktop and run user-provided scripts to run as root?

**Comment by Kommynct** [comment]
Sorry to bump such an old topic but I have found that SKR 

https://aur.archlinux.org/packages/skr-git/
https://github.com/DavRack/skr

Should work well for this in principle.

**Comment by goetzc** [comment]
@Kommynct  I think a better and more mature alternative is [Interception Tools](https://gitlab.com/interception/linux/tools) which was suggested in a [previous](https://github.com/swaywm/sway/issues/719#issuecomment-352227797) comment, using the [k2k](https://github.com/zsugabubus/interception-k2k) plugin.

**Comment by spectras** [comment]
While we're bumping that topic, any updates on how can gaming keyboard effects be implemented on Wayland?

Requirements are as follows. Keyboard management program needs:
- To monitor key presses (sorry about that, but it's hard to give visual feedback to keypresses without the ability to know when keys are pressed).
- To monitor either forground or focused application (lighting profiles are based on active application).
- To run unprivileged.

As long as this cannot be done, there is a whole class of devices that cannot be used fully.
But I did not really follow updates, so maybe I missed some news?

(I could also mention all kinds of business use cases such as screen sharing with remote control, teamviewer, automated GUI testing, robotic process automation software, that just become impossible under Wayland).

**Comment by fluix-dev** [comment]
> While we're bumping that topic, any updates on how can gaming keyboard effects be implemented on Wayland?

Most of this is not in the scope of Wayland (the protocols) or the compositor (Sway) itself:

> To monitor key presses (sorry about that, but it's hard to give visual feedback to keypresses without the ability to know when keys are pressed).

Is this for changing colors? If so, there's already a variety of Linux tools for controlling Razer and other devices and their colours such as the [Openrazer](https://github.com/openrazer/openrazer) drivers, alongside [RazerGenie](https://github.com/z3ntu/RazerGenie) or [Polychromatic](https://github.com/polychromatic/polychromatic) GUI tools. They are perfectly capable, probably through the Openrazer drivers (though I'm not too sure), at listening to key presses and modifying the visual effects.

The main point that people point out with adding some way of monitoring keys is that it's basically a keylogging protocol, which doesn't sound great without some better security implementations first. 

> To monitor either forground or focused application (lighting profiles are based on active application).

The IPC already allows this, you don't really need anything else from Sway.

> (I could also mention all kinds of business use cases such as screen sharing with remote control, teamviewer, automated GUI testing, robotic process automation software, that just become impossible under Wayland).

Look at something like [ydotool](https://github.com/ReimuNotMoe/ydotool) for automated typing and clicking.

**Comment by t184256** [comment]
I think it sounds like a job for remappers like [kmonad](https://github.com/david-janssen/kmonad), not sway.

**Comment by spectras** [comment]
>  Razer and other devices and their colours such as the Openrazer drivers, alongside RazerGenie or Polychromatic GUI tools. They are perfectly capable, probably through the Openrazer drivers (though I'm not too sure), at listening to key presses and modifying the visual effects.

Those need to run as root, and even as kernel module for some features.<br>Except for RazerGenie which relies on a root daemon to bypass wayland device security entirely.

That is indeed a path we could take. Is that really the security model Wayland wants to promote? Take features that could be implemented as regular user-space unprivileged code, and tell developers to make them ring0 and users to give them root permission?

Well I guess that is one way to see it, after all if you distribute the key to everyone, at least when you find your vault empty you can rejoice that is was not broken.

> The main point that people point out with adding some way of monitoring keys is that it's basically a keylogging protocol

The main point is: some use cases require software to know a key was pressed.
- We can implement the use case by ditching security and making everything run as root.
- Or WMs can design a secure protocol, giving control to the user. _“Application foo has requested permission to monitor your keypresses, [insert here whatever additional warning makes sense] — do you agree?”_ seems perfectly reasonable.

By refusing the latter approach, Wayland developers just force people to do the former. The result is a net decrease in overall security.

Also, having it done by the WM would make it possible to have fine control over which keypresses would be visible. Like MS Windows does, where keypresses directed to sensitive applications are not visible to hooks.

---

Anyway, as you see I am less active on this now. We just marked our application as incompatible with Wayland, and the editors we work with simply added "no Wayland" to their system requirements for their own software. Users are fine with that, so that works for us at the moment.

When Wayland is mature enough we can reconsider. Or just run everything as root, if it's really the Wayland's way.

**Comment by fluix-dev** [comment]
@spectras Since my comment I've read a bit more about the state of security in Wayland, so here's some stance/thoughts.

At the moment, as we've both stated, the way for direct event capture is through running something as root. I completely agree with you that this isn't ideal, but it requires at least *some* decision from the user.

I am against adding a protocol to globally grab key events **at this moment** because a protocol for proper security contexts hasn't been established yet (see [here](https://gitlab.freedesktop.org/wayland/weston/-/issues/206) and [here](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/68)). Adding the key grabbing protocol before the security contexts would make unprivileged applications be able to log all keys: arguably *worse* than the current situation. Once a security context protocol has been merged, I think the protocol you're proposing definitely has some merit.

Note: I do not have extensive knowledge of what's discussed here, so take my comments with a grain of salt.

**Comment by kennylevinsen** [comment]
Indeed. Use of root is bad, but this is a net *increase* I'm security from X: rather than everyone and their mother being able to keylog, only root has the ability, and root is omnipotent anyway.

A Wayland protocol will be designed eventually for tasks like keybinding. We will probably still avoid full keylogging. The only reason this does not exist yet is due to other priorities.

There is nothing wrong with falling back on X until Wayland fits your use case.

**Comment by spectras** [comment]
Hey both of you. Thanks for taking the time and energy on this topic.

fluix-dev> I agree, having security contexts makes sense in general, and listening keypresses, being sensitive, could live behind such a mechanism. I am not familiar with wayland development. Is there a way we can help make security contexts happen?

kennylevinsen> my point is the lack of granularity (root or nothing) pushes use cases that used to be user applications to root privileges. Which incurs a decrease in security from X.

In the end, it is the whole class of "user application that needs some special gui permission" that is missing from Wayland, and "make them all root" is hardly a solution. Keypresses are just an example of such a permission.
I think fluix-dev points to the right thing with security contexts, it does look like a clean approach to the more general problem.

**Comment by kennylevinsen** [comment]
> In the end, it is the whole class of "user application that needs some special gui permission" that is missing from Wayland

We have plenty of privileged protocols already. Screen recording, layer shell, clipboard management are all protocols that are not intended for general use and, at some point, will be subject to restriction.

A dedicated key *binding* protocol would be in the same category and require approval, but a omnipotent key logging protocol is unwanted as abuse is unavoidable (just one popular proprietary application misusing it for key bindings is enough) and users would end up unable or unwilling to restrict it, leading to a security situation worse than what we have now: unprivileged keylogging free-for-all.

A key binding protocol is pretty easy to write, but as the primary usecases so far have all been proprietary applications such as discord and zoom that we cannot work on an implementation for (working implementations are a requirement to contribute protocols), no work has To my knowledge been made on this subject. It should be quite a simple protocol though.

**Comment by alex-courtis** [comment]
[dual-function-keys](https://gitlab.com/interception/linux/plugins/dual-function-keys) should fill your requirements.

It's a plugin for [interception tools](https://gitlab.com/interception/linux/tools), which captures and operates directly on udev/evdev devices.

**Comment by t0ster30** [comment]
Try [evdoublebind](https://github.com/exrok/evdoublebind) — it works like xcape, except it uses evdev as a backend (so it can be used both in Sway and i3). Some xkb parts need to be set up though.

**Comment by Kommynct** [comment]
I now prefer keyd for everything related to this, there seems to be many options now for this, this should probably be closed

**Comment by tatsumoto-ren** [comment]
I've been using i3 for more than 10 years. I've tried Sway many times, but couldn't migrate to it, and xcape was one of the reasons. I'm wondering if things are different in 2026. I have this script that runs every time I start my computer:

```bash
#!/bin/bash

# This script is called on startup to remap keys.

# Map right Super key to Menu
xmodmap -e "keysym Super_R = Menu"

# 1. Power saving behaviour of monitors when the computer is not in use
# Disable screen saver blanking and set all the DPMS timeouts to zero.
# https://wiki.archlinux.org/index.php/Display_Power_Management_Signaling
#
# 2. Override LXDE keyboard/mouse settings
# Decrease key repeat delay to 300ms and increase key repeat rate to 50 per second.
#
# Mouse:	xset m accel_mult[/accel_div] [threshold]
# Autorepeat:	xset r rate delay [rate]
# TIP: The q option gives you information on the current settings.
xset \
	s off \
	dpms 0 0 0 \
	m 0 0 \
	r rate 300 50 \
	b off

# Map the caps lock key to super, and map the menu key to right super.
setxkbmap -option terminate:ctrl_alt_bksp,caps:super,altwin:menu_win

# When caps lock is pressed only once, treat it as escape.
killall xcape 2>/dev/null || true
xcape -e 'Super_L=Escape'

# Turn off caps lock if on since there is no longer a key for it.
if xset -q | grep -q "Caps Lock:\s*on"; then
	xdotool key Caps_Lock
fi
```

The idea is that when I press Caps Lock by itself, it acts as Escape, which is handy in programs like Vim for returning from Insert mode to Normal mode. When I press Caps Lock with another key, it acts like the Super key, so I can switch between desktops in i3 by pressing Caps+1, Caps+2, Caps+3, and so on. It's more convenient than pressing Super + number keys because I can do it with one hand. If I had to use the Super key, I'd need to use both hands or position my fingers in an unnatural way.

I didn't invent this setup. I copied it from someone else. So I think other people may be using the same layout. Do you use something similar on Sway? If so, please share your dotfiles, config files, or scripts so I can reproduce this setup. If it's not possible yet, what do you use instead?

I've noticed that in Sway, many things can be configured in the main config file. For example, there's an "output" section to set screen resolution and wallpaper. Is there a feature that would allow defining the settings I have above in the config file as well? That would eliminate the need for an external tool like xcape. If not, how about adding this feature?

**Comment by knezi** [comment]
@tatsumoto-ren 
I use this https://github.com/rvaiya/keyd for the exact use you have. Works great. Even better that it works everywhere - for all users, WMs, tty.

**Comment by tatsumoto-ren** [comment]
@knezi Thank you, this is a good option. However, the config file has to live in /etc/, so you can't simply clone your dotfiles to another machine to replicate an existing configuration. I think this setting should be controlled in `~/.config/sway/config`. For example: 

```
input "type:keyboard" {
  map_once Super_L Escape
}
```

As for other options, so far I have found the following alternatives:

### `xset s off; xset dpms 0 0 0`
 Apparently, aren't needed because sway doesn't put your monitor into standby by default. But `swayidle` can be used if you want to.
### `xset m 0 0` 
  ```
  input "type:pointer" {
      accel_profile flat
      pointer_accel 0
  }
  ```
### `xset r rate 300 50`
  ```
  input "type:keyboard" {
      repeat_delay 300
      repeat_rate  50
  }
  ```
### `xset b off` 
not supported by sway
### `setxkbmap -option terminate:ctrl_alt_bksp,caps:super,altwin:menu_win` 
  ```
  input "type:keyboard" {
    xkb_options terminate:ctrl_alt_bksp,caps:super,altwin:menu_win
  }
  ```
### `xdotool key Caps_Lock` (disable caps if it was enabled)
 can be countered by using keyd, i guess.

**Comment by knezi** [comment]
@tatsumoto-ren I have it in ~/.config and it works fine.
Thinking of it, it's not very safe tho... To change a root config with a regular user.

**Comment by Kommynct** [comment]
i've switched to kanata since i think it works better and should be able to do all of that, via specifying the placement of the config file, highly recommend trying it.

**Comment by tatsumoto-ren** [comment]
The man page for sway says:

> For specifying modifier keys, you can use the XKB modifier names Shift, Lock (for Caps Lock), Control, Mod1 (for Alt), Mod2 (for Num Lock), Mod3 (for XKB modifier Mod3), Mod4 (for the Logo key), and Mod5 (for AltGr). In addition, you can use the aliases Ctrl (for Control), Alt (for Alt), and Super (for the Logo key).

Would it be possible to do the following?

1) Map Caps Lock to Escape (it can be done with `xkb_options caps:escape`)
2) Set Escape as the main modifier key in sway (`set $mod Escape`)

This should eliminate the need to install a separate program like keyd or kanata. Since Escape isn't in the list of available modifiers, would it be okay to submit a PR to add it?

**Comment by kennylevinsen** [comment]
> Escape isn't in the list of available modifiers, would it be okay to submit a PR to add it?

No that wouldn't make sense. But while you cannot make Escape a *modifier* without changing your keyboard layout (i.e., making an XKB file that inherits your current layout and defines Escape as triggering Mod5), that doesn't mean you can't just bind Escape + whatever already. Try that first.

("$mod" is just a random config variable with a random name, it has no special meaning.)

**Comment by tatsumoto-ren** [comment]
@kennylevinsen Wow, this actually works, awesome! It didn't work in i3. So keyd isn't needed after all, that's great.

This configuration worked for me:

```
set $mod Escape
floating_modifier Mod4 # to avoid config parse error
…
input "type:keyboard" {
    …
    xkb_options caps:escape,…
}
```
